ksqldb-engine/src/main/java/io/confluent/ksql/internal/TransientQueryCleanupListener.java

```diff
@@ -0,0 +1,89 @@
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.internal;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import io.confluent.ksql.engine.QueryEventListener;
+import io.confluent.ksql.engine.TransientQueryCleanupService;
+import io.confluent.ksql.engine.TransientQueryCleanupService.TransientQueryStateCleanupTask;
+import io.confluent.ksql.engine.TransientQueryCleanupService.TransientQueryTopicCleanupTask;
+import io.confluent.ksql.metastore.MetaStore;
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.QueryMetadata;
+import io.confluent.ksql.util.TransientQueryMetadata;
+import org.apache.kafka.streams.StreamsConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TransientQueryCleanupListener implements QueryEventListener {
+  private static final Logger log = LoggerFactory.getLogger(TransientQueryCleanupListener.class);
+  private final TransientQueryCleanupService cleanupService;
+  private final ServiceContext serviceContext;
+  private final String stateDir;
+
+  @SuppressFBWarnings(value = "EI_EXPOSE_REP2")
+  public TransientQueryCleanupListener(
+        final TransientQueryCleanupService cleanupService,
+        final ServiceContext serviceContext,
+        final KsqlConfig ksqlConfig) {
+    this.cleanupService = cleanupService;
+    this.serviceContext = serviceContext;
+    this.stateDir = ksqlConfig.getKsqlStreamConfigProps()
+            .getOrDefault(
+                    StreamsConfig.STATE_DIR_CONFIG,
+                    StreamsConfig.configDef()
+                            .defaultValues()
+                            .get(StreamsConfig.STATE_DIR_CONFIG))
+            .toString();
+  }
+
+  @Override
+  public void onCreate(final ServiceContext serviceContext,
+                     final MetaStore metaStore,
+                     final QueryMetadata queryMetadata) {
+    if (queryMetadata instanceof TransientQueryMetadata) {
+      cleanupService.recordTransientQueryOnStart((TransientQueryMetadata) queryMetadata);
+    }
+  }
+
+  @Override
+  public void onClose(
+          final QueryMetadata query
+  ) {
+    if (query instanceof TransientQueryMetadata) {
+      final String applicationId = query.getQueryApplicationId();
+      if (query.hasEverBeenStarted()) {
+        log.info("Cleaning up after query {}", applicationId);
+
+        final TransientQueryTopicCleanupTask topicTask = new TransientQueryTopicCleanupTask(
```

Comment:
This is also done by `QueryCleanupService` in a similar way.  I'm just wondering if we need to do both attempts.  I guess these have the benefit of trying 5 times, but it also does the `findPossiblyLeakedTranientTopics` which also attempts to clean things up based on the current state of what's actually there, which is much more robust in a sense since we don't know if failures appear as failures.  I just wondering, if maybe we need all three types of attempts.  It might not hurt but it's a bit intense.

Reply:
Let's keep the more intense version. Since the exact cause of the leaks is unknown, it doesn't hurt to be more intense here